### ISSUE TITLE
Dexie.Promise not keeping transactions open inside a Web Worker

### ISSUE THREAD

**Issue by dumbmatter** [dfahlander/Dexie.js#483]
I am well aware of the problems with Promises and IndexedDB, and why you need some kind of alternative Promise implementation like Dexie.Promise in most non-Chrome browsers.

However things seem even more complicated inside a Web Worker. [Here is a minimal example (note that the branch is `dexie` not `master` if you want to clone).](https://github.com/dumbmatter/-firefox-indexeddb-promise-worker-test/tree/dexie)

I have some simple code using Dexie.Promise inside an IndexedDB transaction:

    var Dexie = require('dexie');

    var openRequest = indexedDB.open('firefox-indexeddb-promise-worker-test');

    openRequest.onupgradeneeded = function() {
        var db = openRequest.result;
        var store = db.createObjectStore('whatever', {keyPath: 'id'});

        store.put({id: 1});
        store.put({id: 2});
        store.put({id: 3});
    };

    function get(tx, id) {
        return new Dexie.Promise(function (resolve, reject) {
            var req = tx.objectStore('whatever').get(id);
            req.onsuccess = function (e) {
                console.log('got', e.target.result);
                resolve(e.target.result);
            };
            req.onerror = function (e) {
                reject(e.target.error);
            };
        })
    }

    openRequest.onsuccess = function() {
        var db = openRequest.result;

        var tx = db.transaction('whatever');
        tx.oncomplete = function () {
            console.log('tx complete');
        };

        get(tx, 1).then(function () {
            return get(tx, 2);
        }).catch(function (err) {
            console.error(err);
        });
    };

I run this code in a browser, and it works fine. No problem. I run it inside a web worker and I get an error due to the transaction not being active. (If you clone the package, switch to the `dexie` branch, and run `npm start` it will give you links to test both conditions yourself.)

Any idea what's going on, or how to work around this problem?

**Comment by dfahlander** [comment]
Which version of Dexie are you using?

Using Dexie.Promise around native indexedDB code won't help you get around incompabilities between promises and indexeddb since it will also have to schedule a new tick for its callback the same way as native promise does. With Dexie 2.x this is done using a native promise. With 1.x, this is done using other strategies (such as setTimeout(), that is still incompatible with indexedDB transactions).

If you are using Dexie 1.x, the problem could be gone with an upgrade to 2.x.
If you are using 2.x and still have this issue, it means that even Chromium still suffers from incompatibility between promise and indexedDB in workers.

The most secure way to get around these issues is by letting dexie take care of the IDB operations entirely. Of course you can access the native API if you need to, but you must then do that without promises (including Dexie Promise) if it should work across all browsers. (The magic that Dexie.Promise does internally, is to 1) mark the promise with non-standard internal "_lib" flag that scheduls the internal microtask engine, 2) be consistent with this in each async callback.)

**Comment by dfahlander** [comment]
Ok, so I saw you are using Dexie 1.5.1. Try upgrading to dexie@next. When upgraded, you could just use native Promise if you like. Your code will still just work in browsers that have Promise/IndexedDB compatibility anyway.

By the way, I saw you have created an in-memory indexedDB implementation! Would be interesting to try using it in Dexie's unit tests.

**Comment by dumbmatter** [comment]
Thanks for the quick and detailed response! This is going slightly off-topic for Dexie, but I'm just curious how horrible you think something like this is:

    const nativeIDBDatabaseTransaction = IDBDatabase.prototype.transaction;
    const handleTransactionEnd = function () {
        this.txCount -= 1;
        if (this.txCount === 0) {
            self.Promise.idbTransaction = false;
        }
        console.log('tx done', this.name, this.txCount);
    };
    IDBDatabase.prototype.transaction = function (...args) {
        self.Promise.idbTransaction = true;
        if (!this.hasOwnProperty('txCount')) {
            this.txCount = 1;
        } else {
            this.txCount += 1;
        }
        console.log('tx start', this.name, this.txCount);
        const tx = nativeIDBDatabaseTransaction.apply(this, args);
        tx.addEventListener('abort', handleTransactionEnd.bind(this));
        tx.addEventListener('complete', handleTransactionEnd.bind(this));
        tx.addEventListener('error', handleTransactionEnd.bind(this));
        return tx;
    };

Basically I'm doing something like your "_lib" flag automatically when starting a transaction. It feels pretty dirty, but it does seem to work.

Also I'll try to see if I can get Dexie running on fakeIndexedDB. It does have some bugs, but less than some browser implementations :)

**Comment by dfahlander** [comment]
You seem only to reference count the single transaction. Don't understand what it gives?
